### PR TITLE
Control flow API: switch_case

### PR DESCRIPTION
**Pull request by liym27**
- ### Add a control flow API: `switch_case`:
```
switch_case(branch_index, branch_fns, default=None, name=None)
```
> #### `switch_case` is like a C++ `switch-case` statement.
- ### API document
- Chinese Docment: https://github.com/PaddlePaddle/FluidDoc/pull/1629

- English docment:
image

### CODE REVIEW

File: python/paddle/fluid/layers/control_flow.py [hunk 1]
```diff
@@ -2717,6 +2778,154 @@ def _assert_in_rnn_block_(self, method):
...
+                return layers.fill_constant(shape=[3], dtype='int32', value=3)
+
+            main_program = fluid.default_startup_program()
+            startup_program = fluid.default_main_program()
+            with program_guard(main_program, startup_program):
+                index_1 = layers.fill_constant(shape=[1], dtype='int32', value=1)
+                index_2 = layers.fill_constant(shape=[1], dtype='int32', value=2)
+
+                out_1 = layers.switch_case(
+                    branch_index=index_1, branch_fns={1: fn_1,2: fn_2}, default=fn_3)
```
Comment (zhhsplendid) [inline python/paddle/fluid/layers/control_flow.py]: Add a space between "fn_1," and "2:". Or you can try "yapf" the API example code

**Review by zhhsplendid — COMMENTED** [review]
In general, this PR LGTM but the cond API still has to fix something. Let me submit cond and then approve this PR.

File: python/paddle/fluid/layers/control_flow.py (hunk 1 above)
Comment (liym27) [reply python/paddle/fluid/layers/control_flow.py]: Thanks. Done.

File: python/paddle/fluid/layers/control_flow.py [hunk 2]
```diff
@@ -2785,6 +2786,165 @@ def _assert_in_rnn_block_(self, method):
                 method))
 
 
+def switch_case(branch_index, branch_fns, default=None, name=None):
+    '''
+    This operator is like a C++ switch/case statement.
+
+    Args:
+        branch_index(Variable): A Tensor with shape [1] to specify which branch to execute. The data type is ``int32``.
```
Comment (zhhsplendid) [inline python/paddle/fluid/layers/control_flow.py]: Can we support int8, int16, int32, int64 by casting (cast_op)? If we could, I would suggest to not limit the type to int32

File: python/paddle/fluid/layers/control_flow.py [hunk 3]
```diff
@@ -2785,6 +2786,165 @@ def _assert_in_rnn_block_(self, method):
...
+        branch_fns(dict|list|tuple): If it's a list or tuple, the elements in it could be pairs of (int, callable) or
+        simply callables whose actual index will be used as the index of callable. If it's a dict, its key is a python integer and the value is callable.
```
Comment (zhhsplendid) [inline python/paddle/fluid/layers/control_flow.py]: Grammar error: 
1. "simply callables" -> "simple callables". Simple is adjective while simply is adverb.
2. NIT: "and the value is callable." -> "and the value is a callable."

File: python/paddle/fluid/layers/control_flow.py [hunk 4]
```diff
@@ -2785,6 +2786,165 @@ def _assert_in_rnn_block_(self, method):
...
+                            For more information, please refer to :ref:`api_guide_Name`.
+
+    Returns:
+        Variable|list(Variable): Tensors returned by the callable specified by ``branch_index`` in ``branch_fns``,
+        or Tensors returned by ``default`` if ``default`` is not None and no index matches in ``branch_fns``,
+        or Tensors returned by the callable with the max index in ``branch_fns`` if ``default`` is None and no index matches in ``branch_fns``.
+
+    Raises:
+        TypeError: If the type of ``branch_index`` is not Variable.
+        TypeError: If the data type of ``branch_index`` is not int32.
```
Comment (zhhsplendid) [inline python/paddle/fluid/layers/control_flow.py]: As I commented above, suggest to support not only int32

File: python/paddle/fluid/tests/unittests/test_switch_case.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                    1: partial(
+                        layers.fill_constant,
+                        shape=[4, 3],
+                        dtype='int32',
+                        value=1),
+                    2: partial(
+                        fn_1, x=x)
+                })
+            return out
+            #return layers.fill_constant(shape=[4, 2], dtype='int32', value=2)
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_switch_case.py]: Remove useless code？

File: python/paddle/fluid/tests/unittests/test_switch_case.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+                    1: partial(
+                        layers.fill_constant,
+                        shape=[4, 3],
+                        dtype='int32',
+                        value=1),
+                    3: partial(
+                        fn_2, x=3)
+                })
+            return out
+            #return layers.fill_constant(shape=[4, 3], dtype='int32', value=3)
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_switch_case.py]: Also here, remove useless code?

File: python/paddle/fluid/layers/control_flow.py [hunk 7]
```diff
@@ -2785,6 +2786,165 @@ def _assert_in_rnn_block_(self, method):
...
+                "must be {right_value}, but received: {error_value}.".format(
+                    what=what,
+                    arg_name=arg_name,
+                    op_name=op_name,
+                    right_value=right_value,
+                    error_value=error_value))
+
+        if not isinstance(branch_index, Variable):
+            raise TypeError(
+                _error_message("The type", "branch_index", "switch_case",
```
Comment (zhhsplendid) [inline python/paddle/fluid/layers/control_flow.py:2871]: I saw you raise TypeError in _error_message and also raise here, that's not good. 

Two options:
1. Don't raise TypeError in _error_message, but return the error message as a string in _error_message

2. Remove "raise TypeError" here, rename "_error_message" to raise_error_with_message or some other name properly.

File: python/paddle/fluid/layers/control_flow.py (hunk 3 above)
Comment (liym27) [reply python/paddle/fluid/layers/control_flow.py]: Thanks. done

File: python/paddle/fluid/layers/control_flow.py (hunk 7 above)
Comment (liym27) [reply python/paddle/fluid/layers/control_flow.py:2871]: Thanks. Done according to option 1.

File: python/paddle/fluid/tests/unittests/test_switch_case.py (hunk 5 above)
Comment (liym27) [reply python/paddle/fluid/tests/unittests/test_switch_case.py]: Done

File: python/paddle/fluid/tests/unittests/test_switch_case.py (hunk 6 above)
Comment (liym27) [reply python/paddle/fluid/tests/unittests/test_switch_case.py]: Done

File: python/paddle/fluid/layers/control_flow.py (hunk 2 above)
Comment (liym27) [reply python/paddle/fluid/layers/control_flow.py]: Support `uint8`, `int32` and `int64` by casting and op `cast` doesn't support `int8` and `int16`.

File: python/paddle/fluid/layers/control_flow.py (hunk 4 above)
Comment (liym27) [reply python/paddle/fluid/layers/control_flow.py]: Thanks.

File: python/paddle/fluid/tests/unittests/test_switch_case.py [hunk 8]
```diff
@@ -170,14 +169,13 @@ def fn_3():
                         fn_2, x=3)
                 })
             return out
-            #return layers.fill_constant(shape=[4, 3], dtype='int32', value=3)
 
         main_program = Program()
         startup_program = Program()
         with program_guard(main_program, startup_program):
-            index_1 = layers.fill_constant(shape=[1], dtype='int32', value=1)
+            index_1 = layers.data(name="index_1", shape=[1], dtype='uint8')
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_switch_case.py]: Not important.

Unlike fluid.data, layers.data will append -1 as batch size. You feed the array with shape [1] so it didn't have error. But I would suggest to change "layers.data" to "fluid.data" to de-ambigurity

Comment (liym27) [reply python/paddle/fluid/tests/unittests/test_switch_case.py]: Thanks. Done.

**Review by heavengate — APPROVED** [review]
LG API

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Pull request merged**